UDPSocket#recvfrom, #send, and #sendto

tatey opened the pull request:
Hey,

This patch adds support for `recvfrom`, `send`, and `sendto` to `UDPSocket` without any breaking changes. The basic idea is that a Crystal programmer only has to work with `Socket::Addr` which abstracts away the subtle differences between IPv4, IPv6 and the sharp edges of the C APIs.

``` crystal
server = UDPSocket.new
server.bind("localhost", 12345)

client = UDPSocket.new
client.sendto("message equal to buffer".to_slice, server.addr)
message, addr = server.recvfrom(23)

String.new(message) # => "message equal to buffer"
addr.family # => "AF_INET"
addr.ip_address # => "127.0.0.1"
addr.ip_port => 40539

# ...
```

We achieve this by extending `Socket::Addr` to transparently handle the transformation between C SockAddr structs. There's a fair bit of casting between the IPv6 and IPv4 variants that could probably be cleaned up, although some of it feels unavoidable.

This patch also adds the option to set the BROADCAST socket option. Mac OS X will deny sending messages to the broadcast address unless this option is set.

``` crystal
client = UDPSocket.new
client.connect("255.255.255.255", 12349)
client.broadcast = true
```

I'd love to get this patch in to help me with my Crystal implementation of the [LIFX LAN protocol](http://lan.developer.lifx.com). Currently I get around the omission of `recvfrom` by reading in a couple of bytes, determining in if this is a protocol message, and then attempting to parse the rest of the message. This can be avoided by correctly using `recvfrom`. Similar APIs exist in other languages such as [Ruby](http://ruby-doc.org/stdlib-1.9.3/libdoc/socket/rdoc/Socket.html), [Python](https://docs.python.org/3/library/socket.html), and [Go](https://golang.org/pkg/net/#IPConn.ReadFromIP).

Other items to consider:

* Should these functions exist in `IPSocket` and not just `UDPSocket`? The man pages seem indicate it can be used on any type of socket, although I only care about UDP specifically.
* Still need to implement the options that `recvfrom`, `send`, and `sendto` accept.
* More documentation with basic examples

I'm relatively new to these low level APIs and welcome all feedback if you think this patch is useful.

jhass commented on spec/std/socket_spec.cr:
```diff
@@ -202,4 +231,51 @@ describe "UDPSocket" do
+
+  it "sends and receives messages by sendto and recvfrom over IPv4" do
+    server = UDPSocket.new
+    server.bind("localhost", 12347)
```
localhost is not guaranteed to resolve to a v4 address, use a `127.0.0.0/8` address explicitly like you did for v6.

tatey replied:
Good point. Fixed in 23483da3623cb791ca59897a551091528b3b39fc.

jhass commented:
I think adding it to `IPSocket` doesn't make much sense, since use of it will lead to misleading behaviour for TCP sockets, if I'm reading the manpage right, in particular:

> If the socket is connection-mode, dest_addr shall be ignored.

jhass commented on src/socket/socket.cr:
```diff
@@ -32,6 +32,48 @@ class Socket < IO::FileDescriptor
+      @family = case sockaddr.family
+      when LibC::AF_INET then "AF_INET"
+      when LibC::AF_INET6 then "AF_INET6"
+      else raise "Unsupported address family"
```
`ArgumentError`, or perhaps we should make it an enum.

tatey replied:
I'd love to make it use an enum. We already have `Socket::Family` which maps to the underlying C types. The only reason I kept a string was to preserve backwards compatibility with the existing usage of `Socket::Addr`. As long as it was documented as breaking change, would it be OK?

jhass replied:
Yes, breaking everything everywhere if it is for a better solution is still okay.

jhass commented on src/socket/socket.cr:
```diff
@@ -32,6 +32,48 @@ class Socket < IO::FileDescriptor
+      case family
+      when "AF_INET" then to_sockaddrin
+      when "AF_INET6" then to_sockaddrin6
+      else raise "Unsupported address family"
```
Maybe the validation should be in the initializers?

tatey commented:
> I think adding it to IPSocket doesn't make much sense, since use of it will lead to misleading behaviour for TCP sockets, if I'm reading the manpage right, in particular:

Righto.

asterite commented:
I can imagine @waj will want to have a final word on this, so please wait for his answer before merging :-)

tatey commented:
Thanks @asterite. I've just pushed c80dda38fbe5705e58d78a2a21fdf57544645670 which greatly improves the implementation and addresses some of the concerns that @jhass had.

The idea behind this refactor is to remove the ambiguity between sharing `Sock::Addr` with IP and UNIX addresses. This was especially apparent in `IPSocket#addr`, `UDPSocket#sendto`, and `UDPSocket#recvfrom` where you needed to be careful with what types you gave it and what types it gave you.

* `Socket::IPAddr` acts as a nice little wrapper for the C IPv4 and IPv6 structs. Consuming classes don't need to be concerned with the underlying address family. This cleaned up a fair bit of code, but it's obviously a breaking change.
* We correctly raise argument errors if you give `Socket::IPAddr` an unsupported address family. After that we have good data and there's no need to raise exceptions. (Thanks for the feedback here @jhass)
* `Socket::IPAddr` and `Socket::UNIXAddr` can be confident that all their properties are non-nil. They can continue to grow their specialised behaviour without affecting each other.
* The family property is now an enum, although it's not exhaustive.
* I wanted to make `Socket::IPAddr#addrlen` a property that was set at initialization, but the compiler kept flagging it as nillable. That's why there's a funk case statement going on there. Given these structs are now immutable it should be safe to do this. Happy to provide code for clarity.

schmurfy commented:
I don't understand the current UDP socket api, is there something blocking this pull request ?
So far crystal seems nice and I wanted to give it a try for a lightweight server process but I need proper udp support for this :(

schmurfy commented:
I rebased this branch on master and it appears to work for me, I only had to fix a few syntax changes and you also have an error: [sock_addr_in].port is an unsigned number.

ysbaddaden commented:
@schmurfy it merely lacks a proper code review. I glanced quickly and it looks quite good. I'll try to review a bit more when I have some free time.

schmurfy commented:
I will check that tests still passes and link the rebased and fixed branch here, or shall I open a new pull request maybe ?

tatey commented:
Following up with @schmurfy I've had a crack at rebasing and running the tests locally. When I run the tests I get a clang error. I rebased on top of master as of d057006f6088d1fa1fa9f48afef6b9318d21ac8f. Any hints?

```
crystal(test)$ bin/crystal spec spec/std/socket_spec.cr
clang: error: no such file or directory: '/Users/Tate/Code/manastech/crystal/src/ext/libcrystal.a'
Error: execution of command failed with code: 1: `cc -o "/Users/Tate/Code/manastech/crystal/.crystal/crystal-run-spec.tmp" "${@}"  -rdynamic  /Users/Tate/Code/manastech/crystal/src/ext/libcrystal.a -levent -lpcre -lgc -lpthread -ldl`
```

schmurfy commented:
@tatey my branch is available here: https://github.com/schmurfy/crystal/tree/socket if you want to see what I changed, but the only real changes aside from fixes for master is changing [sockaddr_in].port from signed 16 bits to unsigned 16 bits.

I also tried running the tests locally but in my case crystal justs sits there eating most of my CPU without any output (on MAC OS X 10.11)

ysbaddaden commented:
@tatey does `make libcrystal` would help (see #2173)?

asterite commented:
We'll soon take a look at this with @waj, as it seems this is very needed.

tatey commented:
@ysbaddaden It got me further, but still stuck.

```
crystal(socket-sendto-recvfrom*)$ bin/crystal spec spec/std/socket_spec.cr
ld: library not found for -levent
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error: execution of command failed with code: 1: `cc -o "/Users/Tate/Code/manastech/crystal/.crystal/crystal-run-spec.tmp" "${@}"  -rdynamic  /Users/Tate/Code/manastech/crystal/src/ext/libcrystal.a -levent -lpcre -liconv -lgc -lpthread -ldl`
```

I've forked this branch and rebased it against master. I can't actually get it to compile to verify that it still works. I didn't want to force push on top of this branch until we're happy with everything. See https://github.com/crystal-lang/crystal/compare/master...tatey:socket-sendto-recvfrom-rebase-ac46f8c.

ysbaddaden commented:
There is a known issue with some OSX versions (I got it on Maverick, but it usually happens on El Capitan): http://crystal-lang.org/docs/installation/on_mac_osx_using_homebrew.html

tatey commented:
Yeah, I'm running El Capitan. I just tried reinstalling Crystal with the llvm flag and reinstalling the command line tools with no luck. When I get some time I might setup a Linux VM and develop there.

asterite commented:
@tatey Sorry this took so long! Just today we had time to review this.

Your changes were pretty good, we mostly did some renames and cleanups. For example:
* we changed `sento` and `recvfrom` to `send` and `receive`: we don't want to stick with C names, we prefer readable names and a unified naming style. 
* `addr` and `peeraddr` are now `local_address` and `remote_address`
* `IPAddr` => `IPAddress`, we prefer to avoid abbreviations too.

We also removed the `recvfrom(int)` overload because that allocates memory every time you call it. Instead, client code should always provide a buffer (she can allocate it in the heap if necessary).

Maybe missing are overloads for `send` that receive `String`, so the user doesn't have to do `send(string.to_slice)`... @waj thoughts?

tatey commented:
Wow, glad to see this go into master. Thanks for cleaning up and merging. :tada: :+1:

schmurfy commented:
glad to see this merged, but about the naming change from C, are you really sure you want to take that route ?
I already tried that in a slightly different context but my experience is that it's a dangerous pandora's box...

What makes it even worse is when you reuse name that exists in C like send, while I never used it you can use udp socket in pseudo connected mode where send and sendto are two different things and by overwriting one with the other you just create confusion, you might limit the possibilities and you can also create a debugging nightmare when one needs to find later what exactly is the C function called.

That's just my own experience but I think you should really make sure that's what you want to do, especially for an opensource project where other people might/will contribute.

ysbaddaden commented:
Some valid points raised here.

asterite commented:
Yes, we can rename the `send` method that received an IPAddress to `send_to`. It reads weird, though:

```crystal
socket.send_to(message, address)
```

It reads like "send to message". But I don't think it's a big issue. We just used `send` since we support overloads, but `send_to` sounds equally good. I'll check with @waj.

asterite commented:
@schmurfy unless I misunderstood something, do you suggest renaming `send` to `send_to` for the overload that receives an address?

schmurfy commented:
I did not suggest a name but more a general rule, what I was suggesting is to either keep the C syscall name (send, sendto, connect, listen, ...) or at least do not overwrite another name, in that case here is the valid C syscalls for an udp socket to send data:
- send: this will send a datagram to the "connected" address previously set via connect
- sendto: this will send a packet **to** a specific address
- sendmsg: never used that one but it seems to do the same as sendto

In that case you took sendto and renamed it send, that's my main concern, the language does indeed support overload but the real send syscall was not implemented last time I looked at the code, this just increase confusion when learning the language or reading existing code.

I consider C to be what it is: a foundation for almost any language so when I start learning a new one I always try to find my way from the ground to the top based on the POSIX syscalls and most of the time if the language try to be funny and renamed everything I just keep it in the toy section and give up.

I might just be an exception though, no idea :)

asterite commented:
Well, then I don't think it's a good idea to follow C names just because C uses them. I mean, we shouldn't assume that every Crystal programmer used C before.  I mean, in Java it's [send](http://download.java.net/jdk7/archive/b123/docs/api/java/net/DatagramSocket.html#send(java.net.DatagramPacket)) and receives a `DatagramPacket` that might include an address. In Go it's called [WriteTo](https://golang.org/pkg/net/#UDPConn.WriteTo). In C# they have several overloads of [Send](https://msdn.microsoft.com/en-us/library/acf44a1a(v=vs.110).aspx), just like us.

I often see Ruby as a C wrapper with some high-level constructs, dynamisms and metaprogramming. I'm not sure I like that, and I'm not sure it's a good idea. Yesterday I heard in some talk that Matz defines itself as a C programmer (MRI is done in C), so that's probably why Ruby has so many C names in it (puts, gets, sendto, sprintf, etc.). But the idea of a high-level programming language is to abstract over low-level details.

schmurfy commented:
ruby does not really follows the C name (and that's a pain sometimes exactly for the reasons i mentioned before) and I think Matz really is a C programmer since he started on MRI and now appears to mostly work on MRuby: he writes mostly C not ruby ^^
(btw Mruby is what I am currently comparing crystal to, I am already using mruby on multiple projects and loving it, mixing C and ruby is a really pleasant experience with its cleanly defined extension api unlike the MRI chaos).

Assuming that every user used C before is not the topic, being more friendly to the advanced users was more like it but I already explained my thoughts it is up to you which direction you take. I still like the language whether it goes in one direction or the other ;)
